compiler/rustc_parse/src/parser/generics.rs

```diff
@@ -286,7 +304,77 @@ impl<'a> Parser<'a> {
                     },
                 ));
             } else if self.check_type() {
-                where_clause.predicates.push(self.parse_ty_where_predicate()?);
+                // Save the state of the parser before trying to parse a predicate, in case we fail
+                // and encounter something that looks like a misplaced tuple struct body allowing us
+                // to recover from it and fail with a more helpful diagnostic.
+                let parser_snapshot_before_type = self.clone();
```

Comment:
Sadly, I have to clone the parser here even in the happy path incurring a performance cost.
Unsure how to avoid this. As far as I understand, we also do this when parsing `as` casts.
See:
https://github.com/rust-lang/rust/blob/7bbbaabbb6ec93800409478e2af7bc063701604b/compiler/rustc_parse/src/parser/expr.rs#L707

Reply:
I'm not sure the better diagnostics warrrants this complexity, but I guess we'll see what rust perf bot says.

Reply:
You can make the snapshot creation conditional on checking `may_recover` and on `look_ahead(1, |t| t==OpenDelim(Paren))` so that we only pay the cost in those cases. There's also a method in the parser to create the snapshots, I think it's called `create_snapshot_for_diagnostic` (or something along that line) and another to replace `self`, please use those.